[WIP] support no consumer groups in KafkaEx

Gwash3189 opened the pull request:
I noticed that Kafka-Consumer makes the assumption that consumer groups are being used. However, the particular kafka setup i have to interact with does not support consumer groups. 
1. I've added a function that 
   - checks the `kafkaex` configuration for it's `consumer_group`
   - if the consumer group is `:no_consumer_group`
     - adds a `:auto_commit false` to the options list
   - if the consumer group is not `:no_consumer_group`
     - returns an options list containing `worker_name: worker_name`
2. refactors the streaming portion of `consume` as i found it a bit squished / hard to read
   - very happy to remove this portion of the PR if you think it's not needed

Commit:
Implement no consumer group support

Gwash3189 commented:
Sidenote, i found it very difficult to get the tests working / passing for this. 

The tests assume the presence of specific topics, and a local kafka instance to use.
Maybe we could look at some sort of mocking? ( i know it's not normal to mock in elixir, but I feel like not having an entire kafka instance is worth it)

anoskov commented:
Useful changes -- cheers 🍻 @Gwash3189 ! But you know that if auto_commit: false then all messages in topic will be sent each time on connection?
About mock, we tried use https://github.com/artemeff/kafka_ex_mock but it doesn't support consumer offsets. So in our projects we just mocking KafkaEx.produce and then testing EventHandler, without testing KafkaConsumer :(
I will test as soon as the pull request and if everything is ok merge this. Thanks for the contribution!

Gwash3189 commented:
@anoskov

> But you know that if auto_commit: false then all messages in topic will be sent each time on connection?

:hankey: that I did not know. That will prove difficult. Unfortunately, the Kafka infrastructure I'm dealing with at work does not have consumer groups. So I'm a bit stuck :(.

Gwash3189 commented:
@anoskov I had a bit of a play with `KafkaEx.stream` and `offset` today, and `KafkaEx` seems to respect the `offset` value given.

It does not flood the worker module with messages that are before the `offset` value. 

Thoughts?

anoskov commented:
@Gwash3189 Yes you can build workaround offsets. But you should always remember latest message offset, because if you just use KafkaEx.latest_offset function you lose messages that come until worker is down. We decided to just use consumer_groups.

Commit:
coverage for offset server

Gwash3189 commented:
@anoskov I've added an `OffsetServer` behaviour into `KafkaConsumer`. 

This behaviour takes care of the aforementioned offset tracking. 
I've also updated the `README` and added coverage.

👍 / 👎 ?

anoskov commented on README.md:
````diff
@@ -61,6 +61,63 @@ config :kafka_consumer,
+YourModule.OffsetServer.increment("that.one.topic") # :ok
+YourModule.OffsetServer.decrement("that.one.topic") # :ok
+YourModule.OffsetServer.get("that.one.topic") # offset
+```
````
I think you should also support partition number because each partition has its own offset sequence.

Commit:
add doc and spec to start_link

Commit:
support partition

anoskov commented on lib/kafka_consumer/server.ex:
```diff
@@ -44,18 +44,35 @@ defmodule KafkaConsumer.Server do
+          worker_name: worker_name,
+          auto_commit: false,
+          offset: Utils.offset_server().get_latest_offset(topic, partition)
+        ]
```
It's wrong because you should get latest offset from OffsetServer otherwise you lose messages.
Example:
1) You start consumer, latest offset = 5
2) Producer send two messages. Consumer consume it. latest offset = 7
3) Consumer is down.
4) Producer send two messages. Consumer doesn't consume it. latest offset = 9
5) You start consumer and start stream from offset 9 and lose 8 and 9 messages.

Gwash3189 replied:
That's a really good point. I'll add that today /tonight.

anoskov commented on lib/kafka_consumer/offset_server.ex:
```diff
@@ -0,0 +1,145 @@
+      @spec increment(topic) :: noreply
+      def save(topic) do
+        GenServer.cast(make_server_name(topic), :save)
+      end
```
I think it should be fault tolerant. You can try store and get it from DTS table as example. https://github.com/zackehh/stash

Gwash3189 replied:
I'd argue that the user should implement That logic, as different users
would like different things.

Some might want an ETS table, while others want to store the offset in a
database.

Thoughts?

On 12 Sep 2016 9:14 pm, "Andrew Noskov" notifications@github.com wrote:

> In lib/kafka_consumer/offset_server.ex
> https://github.com/anoskov/kafka-consumer/pull/6#discussion_r78353615:
> 
> > +
> > -      @doc """
> > -      decrements the offset
> > -      """
> > -      @spec decrement(topic) :: noreply
> > -      def decrement(topic) do
> > -        GenServer.cast(make_server_name(topic), :decrement)
> > -      end
> >   +
> > -      @doc """
> > -      saves the offset
> > -      """
> > -      @spec increment(topic) :: noreply
> > -      def save(topic) do
> > -        GenServer.cast(make_server_name(topic), :save)
> > -      end
> 
> I think it must be fault tolerant. You can try store and get it from DTS
> table as example. https://github.com/zackehh/stash
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/anoskov/kafka-consumer/pull/6/files/a4689ffc462088bdeb8673c199a0990b13417013#r78353615,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AC5r2OW5tcGv0UASj8YRNEyx3NYQdNPrks5qpTQPgaJpZM4J5qti
> .

anoskov replied:
It's okey i hope. But we must write about it in README.

anoskov commented on lib/kafka_consumer/offset_server.ex:
```diff
@@ -0,0 +1,145 @@
+      defp make_server_name(topic) do
+        String.to_atom(String.replace(topic, ".", "_"))
+      end
+
```
For worker name better use gproc as here https://github.com/anoskov/kafka-consumer/blob/master/lib/kafka_consumer/utils.ex#L58
because atoms are not garbage collected.

anoskov commented on lib/kafka_consumer/offset_server.ex:
```diff
@@ -0,0 +1,145 @@
+      def increment(topic) do
+        GenServer.cast(make_server_name(topic), :increment)
+      end
+
```
Add partition number support in increment/decrement/get/save and make_server_name functions too, please.

Gwash3189 commented:
Hey @anoskov, unfortunately things have shifted at work an I no longer need this functionality. 
I'll close this PR as it's not a priority anymore, but i may return to it because it's fun / and open source is cool.

Thank you for all of your **awesome** feedback and guidance ❤️

anoskov commented:
@Gwash3189 Thanks for your time. I will be glad if you return in the future.
